runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java

```diff
@@ -187,11 +217,152 @@ public void flatMap(T t, Collector<T> collector) throws Exception {
     }
   }
 
-  public void translateGroupByKey(
+  public <K, V> void translateGroupByKey(
           String id,
           RunnerApi.Pipeline pipeline,
           StreamingTranslationContext context) {
-    throw new UnsupportedOperationException();
+
+    //PCollection<KV<K, InputT>> input = context.getInput(transform);
+    RunnerApi.PTransform pTransform =
+            pipeline.getComponents().getTransformsOrThrow(id);
+    String inputPCollectionId =
+            Iterables.getOnlyElement(pTransform.getInputsMap().values());
+    String outputPCollectionId =
+            Iterables.getOnlyElement(pTransform.getOutputsMap().values());
+
+    //@SuppressWarnings("unchecked")
+    //WindowingStrategy<?, BoundedWindow> windowingStrategy =
+    //        (WindowingStrategy<?, BoundedWindow>) input.getWindowingStrategy();
+    RunnerApi.WindowingStrategy windowingStrategyProto =
+            pipeline.getComponents().getWindowingStrategiesOrThrow(
+                    pipeline.getComponents().getPcollectionsOrThrow(
+                            inputPCollectionId).getWindowingStrategyId());
+
+    //KvCoder<K, InputT> inputKvCoder = (KvCoder<K, InputT>) input.getCoder();
+    RunnerApi.Coder inputCoderProto =
+            pipeline.getComponents().getCodersOrThrow(
+                    pipeline.getComponents().getPcollectionsOrThrow(
+                            inputPCollectionId).getCoderId());
+
+    //DataStream<WindowedValue<KV<K, InputT>>> inputDataStream = context.getInputDataStream(input);
+    DataStream<WindowedValue<KV<K, V>>> inputDataStream =
+            context.getDataStreamOrThrow(inputPCollectionId);
+
+    RunnerApi.Coder outputCoderProto =
+            pipeline.getComponents().getCodersOrThrow(
+                    pipeline.getComponents().getPcollectionsOrThrow(
+                            outputPCollectionId).getCoderId());
+
+    RehydratedComponents rehydratedComponents =
+            RehydratedComponents.forComponents(pipeline.getComponents());
+    KvCoder<K, V> inputCoder;
+    try {
+      inputCoder = (KvCoder<K, V>) CoderTranslation.fromProto(
+              inputCoderProto,
+              rehydratedComponents);
+    } catch (IOException e) {
+      throw new IllegalStateException(
+              String.format("Unable to hydrate GroupByKey input coder %s.",
+                      inputCoderProto),
+              e);
+    }
+
+    WindowingStrategy<Object, BoundedWindow> windowingStrategy;
+    try {
+      windowingStrategy =
+              (WindowingStrategy<Object, BoundedWindow>) WindowingStrategyTranslation.fromProto(
```

Comment:
For context, do we already require BoundedWindows in the non-portable runner or is this a new (temporary) restriction?

Reply:
The existing translation doesn't require this, thanks for catching it. I removed the unnecessary cast.